auth0/src/test/java/com/auth0/android/dpop/DPoPTest.kt

```diff
@@ -0,0 +1,518 @@
+        assertThat(result.dpopProof, `is`(notNullValue()))
+
+        val proof = Jwt(result.dpopProof!!)
+        assertThat(proof.decodedPayload["nonce"], `is`(testNonce))
+    }
+
+    @Test
+    public fun `getHeaderData should handle case insensitive DPoP token type`() {
+        whenever(mockKeyStore.hasKeyPair()).thenReturn(true)
+        whenever(mockKeyStore.getKeyPair()).thenReturn(Pair(fakePrivateKey, fakePublicKey))
+
+        val result = DPoP.getHeaderData(
+            "POST", testHttpUrl, testAccessToken, "dpop"
+        )
+
+        assertThat(result.authorizationHeader, `is`("dpop $testAccessToken"))
+        assertThat(result.dpopProof, `is`(notNullValue()))
+    }
+
+    @Test
+    public fun `getHeaderData should return null proof when no key pair exists`() {
+        whenever(mockKeyStore.hasKeyPair()).thenReturn(false)
+
+        val result = DPoP.getHeaderData(
+            "POST", testHttpUrl, testAccessToken, "DPoP"
+        )
+
+        assertThat(result.authorizationHeader, `is`("DPoP $testAccessToken"))
+        assertThat(result.dpopProof, `is`(nullValue()))
+    }
+
+    @Test
+    public fun `clearKeyPair should clear both key pair and nonce`() {
+        val mockNonceResponse = mock<Response>()
+        whenever(mockNonceResponse.headers).thenReturn(
+            Headers.Builder().add("DPoP-Nonce", testNonce).build()
+        )
+        DPoP.storeNonce(mockNonceResponse)
+        assertThat(DPoP.auth0Nonce, `is`(testNonce))
+
+        DPoP.clearKeyPair()
+
+        verify(mockKeyStore).deleteKeyPair()
+        assertThat(DPoP.auth0Nonce, `is`(nullValue()))
+    }
+
+    @Test
+    public fun `clearKeyPair should propagate DPoPException from keyStore`() {
+        whenever(mockKeyStore.deleteKeyPair()).thenThrow(
+            DPoPException.KEY_STORE_ERROR
+        )
+
+        try {
+            DPoP.clearKeyPair()
+            Assert.fail("Expected DPoPException to be thrown")
+        } catch (e: DPoPException) {
+            assertThat(e, `is`(DPoPException.KEY_STORE_ERROR))
+            assertThat(e.message, `is`("Error while accessing the key pair in the keystore."))
+        }
+    }
+
+    @Test
+    public fun `nonce storage should be thread safe`() {
+        val numThreads = 10
+        val numIterations = 100
+        val threads = mutableListOf<Thread>()
+
+        repeat(numThreads) { threadIndex ->
+            threads.add(Thread {
+                repeat(numIterations) { iteration ->
+                    val nonce = "nonce-$threadIndex-$iteration"
+                    val response = mock<Response>()
+                    whenever(response.headers).thenReturn(
+                        Headers.Builder().add("DPoP-Nonce", nonce).build()
+                    )
+                    DPoP.storeNonce(response)
+                }
+            })
+        }
+
+        threads.forEach { it.start() }
+        threads.forEach { it.join() }
+
+        // Should not crash and should have some nonce value
```

Comment:
The problem with non-thread-safe nonce storage is that one could end up with an old (and invalid) nonce stored, while the newer one gets overwritten.